File: solutions/Golang/sort_list_iterative.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	dummy := &ListNode{Next: head}

	length := 0
	curr := head
	for curr != nil {
		length++
		curr = curr.Next
	}

	for size := 1; size < length; size *= 2 {
		prev := dummy
		curr := dummy.Next

		for curr != nil {
			l1 := curr
			l1Tail := l1
			l1Len := 0
			for l1Len < size && l1Tail != nil {
				l1Tail = l1Tail.Next
				l1Len++
			}

			l2 := l1Tail
			l2Len := 0
			for l2Len < size && l2 != nil {
				l2 = l2.Next
				l2Len++
			}

			l1Tail = l1
			for i := 1; i < l1Len; i++ {
				if l1Tail != nil {
					l1Tail = l1Tail.Next
				}
			}
			if l1Tail != nil {
				l1Tail.Next = nil
			}

			prev.Next = merge(l1, l2)
			for prev.Next != nil {
				prev = prev.Next
			}
			curr = l2
		}
	}

	return dummy.Next
}

func merge(l1, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	curr := dummy
	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			curr.Next = l1
			l1 = l1.Next
		} else {
			curr.Next = l2
			l2 = l2.Next
		}
		curr = curr.Next
	}
	if l1 != nil {
		curr.Next = l1
	} else {
		curr.Next = l2
	}
	return dummy.Next
}

func main() {
	head := &ListNode{Val: 4}
	head.Next = &ListNode{Val: 2}
	head.Next.Next = &ListNode{Val: 1}
	head.Next.Next.Next = &ListNode{Val: 3}

	result := sortList(head)
	for result != nil {
		fmt.Print(result.Val, " ")
		result = result.Next
	}
	fmt.Println()
}

File: solutions/C++/sort_list_iterative.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* merge(ListNode* l1, ListNode* l2) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    while (l1 && l2) {
        if (l1->val <= l2->val) {
            curr->next = l1;
            l1 = l1->next;
        } else {
            curr->next = l2;
            l2 = l2->next;
        }
        curr = curr->next;
    }
    curr->next = l1 ? l1 : l2;
    return dummy.next;
}

ListNode* sortList(ListNode* head) {
    if (!head) return head;
    
    ListNode dummy(0);
    dummy.next = head;
    ListNode* dummy_ptr = &dummy;
    
    int length = 0;
    ListNode* curr = head;
    while (curr) {
        length++;
        curr = curr->next;
    }
    
    for (int size = 1; size < length; size *= 2) {
        ListNode* prev = dummy_ptr;
        curr = dummy_ptr->next;
        
        while (curr) {
            ListNode* l1 = curr;
            ListNode* l1_tail = l1;
            int l1_len = 0;
            while (l1_len < size && l1_tail) {
                l1_tail = l1_tail->next;
                l1_len++;
            }
            
            ListNode* l2 = l1_tail;
            int l2_len = 0;
            while (l2_len < size && l2) {
                l2 = l2->next;
                l2_len++;
            }
            
            l1_tail = l1;
            for (int i = 1; i < l1_len; i++) {
                if (l1_tail) l1_tail = l1_tail->next;
            }
            if (l1_tail) l1_tail->next = nullptr;
            
            prev->next = merge(l1, l2);
            while (prev->next) prev = prev->next;
            curr = l2;
        }
    }
    
    return dummy_ptr->next;
}

int main() {
    ListNode* head = new ListNode(4);
    head->next = new ListNode(2);
    head->next->next = new ListNode(1);
    head->next->next->next = new ListNode(3);
    
    ListNode* result = sortList(head);
    while (result) {
        std::cout << result->val << " ";
        result = result->next;
    }
    std::cout << std::endl;
    return 0;
}
